src/Build.UnitTests/ChangeWaves_Tests.cs

```diff
@@ -0,0 +1,219 @@
+                        <Target Name='HelloWorld' Condition="" '$(MSBUILDDISABLEFEATURESFROMVERSION)' == '999.999' and $([MSBuild]::VersionLessThan('" + featureWave + @"', '$(MSBUILDDISABLEFEATURESFROMVERSION)'))"">
+                            <Message Text='Hello World!'/>
+                        </Target>
+                    </Project>";
+
+                TransientTestFile file = env.CreateFile("proj.csproj", projectFile);
+
+                ProjectCollection collection = new ProjectCollection();
+                MockLogger log = new MockLogger();
+                collection.RegisterLogger(log);
+
+                collection.LoadProject(file.Path).Build().ShouldBeTrue();
+                log.AssertLogContains("Hello World!");
+            }
+        }
+
+        [Theory]
+        [InlineData("test")]
+        [InlineData("    ")]
+        [InlineData("")]
+        [InlineData("16-7")]
+        [InlineData("16x7")]
+        [InlineData("16=7")]
+        public void InvalidCallerForIsFeatureEnabledThrows(string waveToCheck)
+        {
+            using (TestEnvironment env = TestEnvironment.Create())
+            {
+                env.SetEnvironmentVariable("MSBUILDDISABLEFEATURESFROMVERSION", "16.8");
+                Shouldly.Should.Throw(() => ChangeWaves.IsChangeWaveEnabled(waveToCheck), typeof(InternalErrorException));
+            }
+        }
+
+        [Theory]
+        [InlineData("test", "16.8")]
+        [InlineData("16_8", "5.7")]
+        [InlineData("16x8", "20.4")]
+        [InlineData("garbage", "18.20")]
+        public void InvalidFormatThrowsWarningAndLeavesFeaturesEnabled(string enabledWave, string featureWave)
+        {
+            using (TestEnvironment env = TestEnvironment.Create())
+            {
+                env.SetEnvironmentVariable("MSBUILDDISABLEFEATURESFROMVERSION", enabledWave);
+                ChangeWaves.IsChangeWaveEnabled(featureWave).ShouldBe(true);
+
+                string projectFile = @"
+                    <Project>
+                        <Target Name='HelloWorld' Condition=""'$(MSBUILDDISABLEFEATURESFROMVERSION)' == '" + ChangeWaves.EnableAllFeatures + @"' and $([MSBuild]::VersionLessThan('" + featureWave + @"', '$(MSBUILDDISABLEFEATURESFROMVERSION)'))"">
+                            <Message Text='Hello World!'/>
+                        </Target>
+                    </Project>";
+
+                TransientTestFile file = env.CreateFile("proj.csproj", projectFile);
+
+                ProjectCollection collection = new ProjectCollection();
+                MockLogger log = new MockLogger();
+                collection.RegisterLogger(log);
+
+                Project p = collection.LoadProject(file.Path);
+                p.Build().ShouldBeTrue();
+
+                log.AssertLogContains("invalid format");
+                log.AssertLogContains("Hello World!");
+            }
+        }
+
+        [Theory]
+        [InlineData("0.8")]
+        [InlineData("203.45")]
```

Comment:
Should we decide to enable or disable the features based on whether the number is above or below the current waves?

Reply:
If `MSBuildDisableFeaturesFromVersion` is set to something out of bounds, we default to the lowest wave being set. Effectively disabling all waves/features behind those waves.

Reply:
I'm wondering if it makes sense to take into account where the value they specified is relative to other change waves, so if 16.8 and 16.10 are in rotation, and the user specifies 16.5, treat it as "all"; if they select 16.11, treat it as "none"; and if they select 16.9, treat it as 16.10. What do you think?

Reply:
I'm somewhat against having a technically-invalid value such as 16.9 being treated as 16.10. It seems like overcomplicating things by adding another rule.

I would consider having a "clamp" to the nearest wave when the user sets a wave completely out of bounds if it makes more sense to work that way. The way it works now (with your versions):
16.5->16.8
16.11->16.8
And the clamp would work like so:
16.5->16.8
16.11->16.10

Reply:
On a related note, If there's a realistic scenario for a customer wanting the highest wave at all times, we can have a special value like "highest" that will always clamp to the highest, and same for lowest. This would avoid producing warnings and give them an opt out, but may be against the spirit of this feature.

Reply:
I actually think of it as having one fewer rule. Instead of separating invalid versions from valid ones, it's just:
If your version is valid, take all features after the provided version.